pkg/api/resource_types.go

```diff
@@ -62,28 +79,199 @@ func (d *Resource) GetDeletionTimestamp() *metav1.Time {
+// into resource manifest (map[string]interface{}).
+func DecodeManifest(manifest datatypes.JSONMap) (map[string]interface{}, error) {
+	if len(manifest) == 0 {
+		return nil, nil
+	}
+
+	evt, err := JSONMAPToCloudEvent(manifest)
 	if err != nil {
-		return nil, errors.GeneralError("Unable to marshal resource jsonmap status: %s", err)
+		return nil, fmt.Errorf("failed to convert resource manifest to cloudevent: %v", err)
 	}
-	resourceStatus := &ResourceStatus{}
-	if err := json.Unmarshal(resourceStatusJSON, resourceStatus); err != nil {
-		return nil, errors.GeneralError("Unable to unmarshal resource status: %s", err)
+
+	eventPayload := &workpayload.Manifest{}
+	if err := evt.DataAs(eventPayload); err != nil {
+		return nil, fmt.Errorf("failed to decode cloudevent payload as resource manifest: %v", err)
 	}
 
-	return resourceStatus, nil
+	return eventPayload.Manifest.Object, nil
 }
 
-func ResourceStatusToJSONMap(status *ResourceStatus) (datatypes.JSONMap, error) {
-	resourceStatusJSON, err := json.Marshal(status)
+// DecodeManifestBundle converts a CloudEvent JSONMap representation of a list of resource manifest
+// into a list of resource manifest (map[string]interface{}).
+func DecodeManifestBundle(manifest datatypes.JSONMap) ([]map[string]interface{}, error) {
+	if len(manifest) == 0 {
+		return nil, nil
+	}
+
+	evt, err := JSONMAPToCloudEvent(manifest)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert resource manifest to cloudevent: %v", err)
+	}
+
+	eventPayload := &workpayload.ManifestBundle{}
+	if err := evt.DataAs(eventPayload); err != nil {
+		return nil, fmt.Errorf("failed to decode cloudevent payload as resource manifest bundle: %v", err)
+	}
+
+	manifests := make([]map[string]interface{}, 0, len(eventPayload.Manifests))
+	for _, m := range eventPayload.Manifests {
+		if len(m.Raw) == 0 {
+			return nil, fmt.Errorf("manifest in bundle is empty")
+		}
+		// unmarshal the raw JSON into the object
+		obj := &map[string]interface{}{}
+		if err := json.Unmarshal(m.Raw, obj); err != nil {
+			return nil, fmt.Errorf("failed to unmarshal manifest in bundle: %v", err)
+		}
+		manifests = append(manifests, *obj)
+	}
+
+	return manifests, nil
+}
+
+// DecodeManifest converts a CloudEvent JSONMap representation of a resource status
+// into resource status (map[string]interface{}).
+func DecodeStatus(status datatypes.JSONMap) (map[string]interface{}, error) {
+	if len(status) == 0 {
+		return nil, nil
+	}
+
+	evt, err := JSONMAPToCloudEvent(status)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert resource status to cloudevent: %v", err)
+	}
+
+	evtExtensions := evt.Extensions()
+	resourceVersionInt := int64(0)
+	resourceVersion, err := cloudeventstypes.ToString(evtExtensions[cetypes.ExtensionResourceVersion])
+	if err != nil {
+		resourceVersionIntVal, err := cloudeventstypes.ToInteger(evtExtensions[cetypes.ExtensionResourceVersion])
```

Comment:
only use `ToInteger`?

Reply:
right, we now ensure the resource version is integer.
Updated all the occurrences.